Refactor builtin HUD

cx384 opened the pull request:
## Goal of the PR
- Refactor the builtin code for HUD elements, such that it can handle other HUD elements besides statusbars.
- It is needed for #14071 #14273 #14321 and potential future elements, for example for the chat or infotext.
- ~~(The statusbars behave the same as in the old version, the only difference is that `hud_replace_builtin` does not default `definition.item` anymore, but this is not documented in`lua_api.md` so it should be fine. However, I can re-add it if someone thinks it is needed.)~~ readded

## Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
2.2 Internal code refactoring

## To do

This PR is Ready for Review.

## How to test

- Start devtest, jump into water to test the breathbar, and hurt yourself to test the healthbar.
- Read the code notice the refactoring.
- Test `hud_replace_builtin`
```lua
minetest.hud_replace_builtin("breath", {
	type = "statbar",
	position = {x = 0.5, y = 0.5},
	text = "bubble.png",
	text2 = "bubble_gone.png",
	number = core.PLAYER_MAX_BREATH_DEFAULT * 2,
	item = core.PLAYER_MAX_BREATH_DEFAULT * 2,
	direction = 0,
	size = {x = 24, y = 24},
	offset = {x = 25, y= -(48 + 24 + 16)}
})
```
- Test flags
```lua
minetest.register_chatcommand("hudtoggleflag", {
	description = "Toggles a hud flag.",
	params = "<flag>",
	func = function(name, params)
		local player = minetest.get_player_by_name(name)
		if not player then
			return false, "No player."
		end

		local flags = player:hud_get_flags()
		if flags[params] == nil then
			return false, "Unknown hud flag."
		end

		flags[params] = not flags[params]
		player:hud_set_flags(flags)
		return true, "Flag \"".. params .."\" set to ".. tostring(flags[params]) .. "."
	end
})
```
`/hudtoggleflag breathbar` `/hudtoggleflag healthbar`
- Also test other builtin elements in case new ones get added by other PRs.

cx384 committed:
Refactor builtin HUD

cx384 committed:
Re-add minimap

cx384 commented:
Rebased.
This PR now also fixes a warning you get because `hud_elem_type` has been renamed to `type`.
I caused this with #14071, because the PR was made before the renaming, and I missed to change it. Sorry about this.

cx384 commented:
Rebased again.

sfan5 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,238 @@
+function core.hud_replace_builtin(elem_name, definition)
+	if type(definition) ~= "table" or
+			(definition.type or definition.hud_elem_type) ~= "statbar" then
+		return false
```
should be an `assert` imo

cx384 replied:
Oh, I also forgot to remove the statbar check when I rebased and the minimap element got added.

cx384 replied:
I replaced it with an assert, but this is now a change in the behavior of `minetest.hud_replace_builtin` since the unrefactored code just returned `false` in this case.

sfan5 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,238 @@
+			return show_breathbar
+		else
+			-- Only if the element does not already exist the breath amount is considered.
+			return show_breathbar and player:get_breath() < player:get_properties().breath_max
```
forgot to reference `breath` `breath_max` variables above?

cx384 replied:
Fixed.

cx384 committed:
Reuse breath and breath_max

cx384 committed:
Use assert and fix statbar check

cx384 commented:
Thanks for reviewing it.
With the refactoring, I didn't want to change any behavior (even if it is not documented in the API).
However, I applied your `assert` suggestion now, so `minetest.hud_replace_builtin` behaves a little differently, but I guess this is good.

I also rebased to make sure that it still works, since it is almost 2 months old.

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+		local player = core.get_player_by_name(playername)
+		local id = player_hud_ids[elem_name]
+		if player and id then
+			player:hud_remove(id)
```
This should also reset the ID in `player_hud_ids` to nil

Otherwise doing something like this in a mod deletes the HUD element instead of replacing it:

```lua

core.after(6, function()
	core.hud_replace_builtin("minimap", {
		type = "minimap",
		position = { x = 0, y = 0 },
		alignment = {x = -1, y = 1},
		offset = {x = -10, y = 10},
		size = {x = 256, y = 256},
	})
end)
```

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+						player:hud_remove(id)
+					end
+				end, player_name, id)
+				hud_ids[player_name].breath = nil
```
This means that there will be two breathbars if you lose some breath again in this one second

cx384 replied:
Yes it does, as it was before the refactoring.
Should I add a mechanism to not show it multiple times?

grorp replied:
That would be nice, but since this is a refactoring, adding a comment describing the known bug would be enough too.

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+	local name = player:get_player_name()
+	if name == "" then
+		return
+	end
```
How, possibly, would a player with the name `""` leave?

cx384 replied:
I took this from the old version, I guess it happens when the player somehow does not exist anymore.
Would you prefer a `minetest.is_player(player)` check?

grorp replied:
This function is only called as a callback given to `register_on_leaveplayer`. If Minetest deleted or invalidated objects before calling `register_on_leaveplayer` callbacks, we'd have a serious problem and many mods would be broken. What I'd prefer is removing the check completely, as it gives the false impression of being necessary.

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+end
+
+-- Returns true if successful, otherwise false,
+-- but currently the return value is not part of the lua API.
```
```suggestion
-- but currently the return value is not part of the documented Lua API.
```

cx384 committed:
Improve comments

cx384 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+					if player then
+						player:hud_remove(id)
+					end
+				end, player_name, id)
```
Inherited from the old version, but I changed it now.

cx384 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+	end
+end
+
+local function player_event_handler(player, eventname)
```
I don't know, this was also the case for the old version.
Should I remove the return statements?

grorp replied:
Yes

cx384 committed:
Apply suggestions

cx384 committed:
Reuse update_hud to avoid duplication

cx384 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,235 @@
+		end
+		for _, elem_name in ipairs(to_update) do
+			update_element(player, player_hud_ids, elem_name, flags)
+		end
```
`update_hud` is already meant to do this but for all elements, so I modified it such that it can be reused.

cx384 committed:
Apply missed suggestion

cx384 committed:
Fix l

cx384 committed:
Remove player_event_handler bool returns

cx384 committed:
Remove player check

cx384 committed:
Document bug

cx384 commented:
I resolved everything.
The breathbar bug stays as is for now. Maybe someone wants to fix it in a separate PR.
I tested it again, so I hope everything still works fine.

cx384 committed:
Remove variable

grorp commented:
Hey @rollerozxa, can you do a zipgrep for `hud_replace_builtin` please?

rollerozxa commented:
sure, here: https://content.minetest.net/zipgrep/cf8a1325-8075-44f6-a32c-f10e6ad93f83/

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,225 @@
+	show_elem = function(player, flags, id)
+		local breath     = player:get_breath()
+		local breath_max = player:get_properties().breath_max
+		local show_breathbar = flags.breathbar and enable_damage and player:get_armor_groups().immortal ~= 1
```
line too long (>95), should be wrapped

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,225 @@
+	},
+	show_elem = function(player, flags, id)
+		local breath     = player:get_breath()
+		local breath_max = player:get_properties().breath_max
```
I suggest using a local variable like this for the conditions below:
```lua
local breath_relevant = player:get_breath() < player:get_properties().breath_max
```

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,225 @@
+			if breath == breath_max then
+				local player_name = player:get_player_name()
+				-- The breathbar stays for some time and then gets removed.
+				-- NOTE: This can causes a bug where multiple breathbars are shown at the same time.
```
```suggestion
				-- NOTE: This can cause a bug where multiple breathbars are shown at the same time.
```

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,225 @@
+				hud_ids[player_name].breath = nil
+			end
+
+			return show_breathbar
```
This should mean that the breathbar is immediately recreated (= duplicated) after the ID is set to nil by the `if breath == breath_max` code section above. The only reason this doesn't happen is that `update_element` caches the ID and doesn't notice it has become nil.

While this happens to work, maybe by accident, it's confusing/misleading.

grorp replied:
I suppose the cleanest solution to this problem would be moving the whole "remove after 1 second if breath == breath_max" code to `hud_change` instead of `show_elem`. Using a getter function for this seems wrong anyway.

cx384 replied:
This was actually on propose to behave like the old code, but I made it more clear now, and fixed the duplicate breath bar bug.

grorp commented:
> (The statusbars behave the same as in the old version, the only difference is that `hud_replace_builtin` does not default `definition.item` anymore, but this is not documented in`lua_api.md` so it should be fine. However, I can re-add it if someone thinks it is needed.)

Looking through the zipgrep kindly provided by rollerozxa, I can see at least one package relying on this default value (https://content.minetest.net/packages/epCode/what_were_you_expecting/). You have to re-add it for backwards compatibility (and it would be good to document and/or deprecate it).

cx384 committed:
Update builtin/game/hud.lua

Fix grammar

Co-authored-by: grorp <gregor.parzefall@posteo.de>

cx384 committed:
Update builtin/game/hud.lua

Fix grammar2

Co-authored-by: grorp <gregor.parzefall@posteo.de>

cx384 committed:
Change comment

Co-authored-by: grorp <gregor.parzefall@posteo.de>

cx384 committed:
Apply suggestions and fix duplicate breathbar

cx384 committed:
Remove deprecated flags change

cx384 commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,225 @@
+
+	if def.hud_change then
+		def.hud_change(player, id)
+	end
```
> it should be easy enough to fix by providing elem_def as a function returning a table.

This is not possible, since `elem_def` is set by `hud_replace_builtin`.

Instead, I added a `init_change` function. Now it looks like code duplication, but at least it is faster.

cx384 commented on builtin/game/statbars.lua:
```diff
@@ -1,217 +0,0 @@
-		hud_ids[name] = {}
-		-- flags are not transmitted to client on connect, we need to make sure
-		-- our current flags are transmitted by sending them actively
-		player:hud_set_flags(flags)
```
Ok, I removed it now, since you said that it can safely be removed.

cx384 commented:
I resolved everything again, and re-added the undocumented `elem_def.item` behavior.

Also, I fixed the duplicate breatbbar bug.

grorp committed:
Fix broken delayed removal, move removal code out of getter function

grorp committed:
More consistent naming, reorder code in update_def to actually fix backwards compat

scaleToHudMax uses elem_def.item, so it must be set first to avoid a Lua error with mods
that rely on the old behavior

Remove "deprecated undocumented behavior" comment since it doesn't make much sense -
how are modders supposed to know that something is deprecated if it's not documented?

grorp committed:
Update code style according to https://dev.minetest.net/Lua_code_style_guidelines

Specifically:
- "Functions and variables should be named in lowercase_underscore_style"
- "When breaking around a binary operator you should break after the operator."

grorp committed:
Consistent order for HUD definition members

grorp committed:
Don't update the minimap element on "properties_changed" events

grorp committed:
Fix docs?

grorp commented:
For reference: https://github.com/cx384/minetest/pull/2

cx384 committed:
Merge pull request #2 from grorp/builtin_hud_refactor

Suggestions by grorp

cx384 commented:
> For reference: [cx384#2](https://github.com/cx384/minetest/pull/2)

Seams good, merged. Sorry for being a little late. 
(Maybe I should have refactored more and not take too many things from the old code.)

grorp commented on builtin/game/hud.lua:
```diff
@@ -0,0 +1,261 @@
+	show_elem = function(player, flags)
+		-- Don't add a minimap for clients which already have it hardcoded in C++.
+		return flags.minimap and
+				minetest.get_player_information(player:get_player_name()).protocol_version >= 44
```
```suggestion
				core.get_player_information(player:get_player_name()).protocol_version >= 44
```

grorp committed:
Fix `minetest` being used instead of `core`

grorp reviewed (approved):
This PR looks OK to me now.

@sfan5 since there have been a lot of changes, you might want to have a look at this again.

sfan5 commented:
> @sfan5 since there have been a lot of changes, you might want to have a look at this again.

I trust your judgement.
